unkeyed/unkey
# feat(web): add budget spend management

## What does this PR do?



Fixes #1223 

I've added a Loom so that you can see the feature of interacting with budgets in action:
https://www.loom.com/share/986fc21fb8aa4ed6910494838826780d?sid=0a5fe5cd-c94b-4655-85ba-37afaf085ccc

The email that users will receive is the following:

<img src="https://github.com/unkeyed/unkey/assets/51177379/9efc0f6f-8cb6-4c0e-832f-e78b3cdf03a1)" width="fit" height="400">



## Type of change




## How should this be tested?



Requirements for testing:
- Tinybird (to retrieve active keys and verification for usage calculation). 
- Stripe (basically for creating a paid workspace). Believe you could use dummy env vars for this so that `stripeSchema` doesn't throw when you create a paid workspace.
- App
- Worfklows (trigger.dev)

- **Test A**: Handle a Budget.
  1. Go to the Settings section and head to the Billings tab
  2. You should be able to create a budget with the "Create Budget" button on the "Budgets section". There you can pass the name of the budget, the amount an additional emails to notify whenever the spending reaches the limit.
  3. Once created, you are able to edit or delete your bucket.
- **Test B**: Receive a spending limit notification
   1. This test requires you spin the "workflows" trigger app. 
   2. Once the job is running (it's running on a 10min interval basis), try to exceed the specified limit directly on Tinybird and wait for the queue to check on your workspace budget.
   3. After the job catched your spending exceeded you should: i) receive an email about the usage, ii) have a `lastReachedAt` property set on your budget (see db) and iii) should not receive more emails about that budget for the rest of the month cycle.

## Checklist



### Required


### Appreciated
